engine/collection/compliance/core.go

```diff
@@ -108,6 +117,20 @@ func (c *Core) OnBlockProposal(originID flow.Identifier, proposal *messages.Clus
 		return fmt.Errorf("could not check proposal: %w", err)
 	}
 
+	// ignore proposals which are too far ahead of our local finalized state
+	// instead, rely on sync engine to catch up finalization more effectively, and avoid
+	// large subtree of blocks to be cached.
+	final, err := c.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("could not get latest finalized header: %w", err)
+	}
```

Comment:
I would suggest using [FinalizedHeaderCache](https://github.com/onflow/flow-go/blob/a2c545d78087b26f97ccea09022f61c3b124d5bb/engine/common/synchronization/finalized_snapshot.go#L32) instead, so we can avoid accessing the state every time.

Reply:
I don't think the `FinalizedHeaderCache` is a great way to cache this data, which is why I avoided using it here. Some reasoning below:

We already have resource-specific caches, including for [headers](https://github.com/onflow/flow-go/blob/2366de82cbc5e6c68eadbadbe246784c16fc46aa/storage/badger/headers.go#L22). We are all but guaranteed to always find the latest finalized header in this cache without touching the database.

In addition to retrieving the header, `state.Final` will query for the `finalized_height` pointer and the corresponding `height->block_id` lookup from the database. These lookups would benefit from something like `FinalizedHeaderCache`. However, if we are going to cache these values, why not cache them in the protocol state itself? Then all protocol state users get the performance improvement, without needing to incorporate another component.

Reply:
Ah I haven't looked into what has been implemented for caching in the protocol state itself. It probably makes more sense to do that.

Reply:
Added an issue for it: https://github.com/dapperlabs/flow-go/issues/6224